src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,218 @@
+            cartEntries = cart.getEntries();
+            CartEntry cartEntry = cartEntries.stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().orElse(null);
+
+            if (cartEntry != null) {
+                cartEntry.setQuantity(cartEntry.getQuantity() + quantity);
+                cartEntry.setCart(cart);
+            } else {
+                CartEntry newCartEntry = new CartEntry();
+                newCartEntry.setProductName(productName);
+                newCartEntry.setQuantity(quantity);
+                newCartEntry.setCart(cart);
+                cartEntries.add(newCartEntry);
+            }
         }
 
-        cart.setUser(user);
-
         for (int i = 0; i < quantity; i++) {
-            productList.add(product);
-
-            productTotal += product.getProductPrice();
+            productTotal += productDTO.getPrice();
         }
 
-        cart.setProductList(productList);
+        cart.setEntries(cartEntries);
         cart.setTotal(cart.getTotal() + productTotal);
+        cart.setUserId(user.getId());
 
-        cartRepository.save(cart);
-
-        return CartMapper.cartEntityToDto(cart);
+        ;
 
+        return cartEntityToDto(cartRepository.save(cart));
     }
 
     public CartDTO updateProductFromCart(String email, String productName, int quantity)
             throws InvalidInputDataException, EntityNotFoundException {
+        log.info("removeProductFromCart - process started");
 
-        Product product;
+        ProductDTO productDTO;
         try {
-            product = ProductMapper.DtoToProductEntity(productServiceClient.callGetProductByProductName(productName));
+            productDTO = productServiceClient.callGetProductByProductName(productName);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Product: " + productName + " not found in the Database!");
         }
 
-        User user;
-        try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
-        } catch (HttpClientErrorException.NotFound e) {
-            throw new EntityNotFoundException("UserName: " + email + " not found in the Database!");
+        if (productDTO.getStock() == 0) {
+            throw new InvalidInputDataException("We are sorry, but currently: " + productName + " is out of order!");
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
-
-        Long productsInCart = productList.stream().filter(p -> p.getProductName().equals(productName)).count();
-
-
 
-        if (productsInCart == 0) {
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
+        if (productDTO.getStock() < quantity) {
+            throw new InvalidInputDataException("You can not remove more than: " + productDTO.getStock() + " "
```

Comment:
the message is a little bit misleading. the user does not remove a product but is actually adding products to the cart

Reply:
Done!